gloo/transport/tcp/tls/pair.cc

```diff
@@ -0,0 +1,386 @@
+  // means some other thread is trying to mutate this pair's state,
+  // which in turn might require calling into (and locking) the
+  // underlying device (for example, when the pair transitions to the
+  // CLOSED state). To avoid deadlocks, attempt to lock the pair and
+  // skip handling the events until the next tick if the lock cannot
+  // be acquired.
+  std::unique_lock<std::mutex> lock(m_, std::try_to_lock);
+  if (!lock) {
+    return;
+  }
+
+  // State must be <= CONNECTED.
+  // If state is CLOSED; this function will NOT be called. Refer to
+  // Pair::changeState and Device::unregisterDescriptor for more info.
+  GLOO_ENFORCE_LE(state_, CONNECTED);
+
+  // Exception must not be set.
+  // If exception is set, state must advance to CLOSED state.
+  GLOO_ENFORCE(ex_ == nullptr);
+
+  if (state_ == CONNECTED) {
+    if (!is_ssl_connected_) {
+      if (!is_client_) {
+        int events = 0;
+        if ((events = handshake())) {
+          device_->registerDescriptor(fd_, events, this);
+        }
+      }
+    } else {
+      if (events & EPOLLOUT) {
+        GLOO_ENFORCE(
+            !tx_.empty(), "tx_ cannot be empty because EPOLLOUT happened");
+        while (!tx_.empty()) {
+          auto& op = tx_.front();
+          if (!write(op)) {
+            // Write did not complete; wait for epoll.
+            break;
+          }
+          // Write completed; remove from queue.
+          tx_.pop_front();
+        }
+        // If there is nothing to transmit; remove EPOLLOUT.
+        if (tx_.empty()) {
+          device_->registerDescriptor(fd_, EPOLLIN, this);
+        }
+      }
+      if (events & EPOLLIN) {
+        while (read()) {
+          // Keep going
+        }
+      }
+    }
+    return;
+  }
+
+  if (state_ == LISTENING) {
+    handleListening();
+    return;
+  }
+
+  if (state_ == CONNECTING) {
+    handleConnecting();
+    return;
+  }
+
+  GLOO_ENFORCE(false, "Unexpected state: ", state_);
+}
+
+void Pair::changeState(Pair::state nextState) noexcept {
+  if (nextState == CLOSED && is_ssl_connected_) {
+    SSL_shutdown(ssl_);
+    SSL_free(ssl_);
+    is_ssl_connected_ = false;
+  }
+  ::gloo::transport::tcp::Pair::changeState(nextState);
+}
+
+void Pair::waitUntilSSLConnected(
+    std::unique_lock<std::mutex>& lock,
+    bool useTimeout) {
+  auto pred = [&] {
+    throwIfException();
+    return is_ssl_connected_;
+  };
+  waitUntil(pred, lock, useTimeout);
+}
+
+void Pair::waitUntilConnected(std::unique_lock<std::mutex> &lock,
+                              bool useTimeout) {
+  ::gloo::transport::tcp::Pair::waitUntilConnected(lock, useTimeout);
+
+  if (!is_ssl_connected_) {
+    if (is_client_) {
+      int events;
+      while ((events = handshake())) {
+        int r;
+        struct pollfd pfd;
+        pfd.fd = fd_;
+        pfd.events = events;
+        do {
+          r = poll(&pfd, 1, 100);
+        } while (r == 0);
+        GLOO_ENFORCE(r == 1, "poll return %d error events: %d errno %d %s\n",
```

Comment:
is it true that there is only one socket so that `r` should also equal to 1? BTW, can we add a link to poll's return value explanation in comments?

Reply:
It can return negative value according to the doc

Reply:
I added a comment